pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,232 @@
+	tc := &LegacySATokenCleaner{
+		client:                       cl,
+		saLister:                     saInformer.Lister(),
+		saInformerSynced:             saInformer.Informer().HasSynced,
+		secretLister:                 secretInformer.Lister(),
+		secretInformerSynced:         secretInformer.Informer().HasSynced,
+		configMapLister:              configMapInformer.Lister(),
+		configMapInformerSynced:      configMapInformer.Informer().HasSynced,
+		podLister:                    podInformer.Lister(),
+		podInformerSynced:            podInformer.Informer().HasSynced,
+		legacySATokenCleanUpPeriod:   options.LegacySATokenCleanUpPeriod,
+		legacySATokenCleanUpInterval: options.LegacySATokenCleanUpInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.configMapInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.legacySATokenCleanUpInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	secretList, err := tc.secretLister.Secrets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "Getting secret list from secret cache")
+		return
+	}
+	for _, secret := range secretList {
+		sa, saErr := tc.getServiceAccount(secret)
+		switch {
+		case saErr != nil:
+			logger.Error(saErr, "Getting service account")
+		case sa == nil:
+			logger.V(4).Info("Service account does not exist", "secret", klog.KRef(secret.Namespace, secret.Name))
+		case tc.shouldCleanToken(logger, secret, sa):
+			logger.Info("Auto-generated service account token is not used for more than 365 * 24 hours, deleting token", "secret", klog.KRef(secret.Namespace, secret.Name))
+			if err := tc.client.CoreV1().Secrets(secret.Namespace).Delete(context.TODO(), secret.Name, metav1.DeleteOptions{}); err != nil {
+				logger.Error(err, "Deleting legacy service account token", "secret", klog.KRef(secret.Namespace, secret.Name), "service account", sa.Name)
+			}
+		}
+	}
+}
+
+func (tc *LegacySATokenCleaner) getServiceAccount(secret *v1.Secret) (*v1.ServiceAccount, error) {
+	saName := secret.Annotations[v1.ServiceAccountNameKey]
+	saUID := types.UID(secret.Annotations[v1.ServiceAccountUIDKey])
+	sa, err := tc.saLister.ServiceAccounts(secret.Namespace).Get(saName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
```

Comment:
the below might be cleaner in terms of possible extensions to the function and allows you to omit the `if sa != nil` condition further on
```go
if err != nil {
    if apierrrors.IsNotFound(err) {
        return nil, nil
    }
    return err
}
```
I'm not sure you can get a different than NotFound error from the cache, though

Reply:
I am not sure the `if sa != nil` can be omitted. If the service account is deleted and created again (with the same namespace and name), the uid should be different than the original one, right?

Reply:
Oh maybe you mean just remove the `if sa != nil` condition, but not the inner part?

Reply:
Done.